README.md
# create-key

Terminal assistant to generate or read cryptographic keys. This utility was created out of frustration, as more time was spent Googling _how to use openssl_ rather than getting things done. This utility is originally aimed to be used via `npx`, but it can also be used in NodeJS programs. It is not as powerful as openssl as it only support RSA and ECDSA ciphers. Thanks to `npx`, there is no need to install this utility, just make sure that `npm` > 5.2 is installed and then run:

```
npx create-keys
```

This command starts a terminal questionnaire that helps building the keys you need. 

Currently, this package only supports the following:
- Ciphers: RSA or ECDSA
- Encoding: PCKS8 for private keys, PCKS1 for RSA public keys and SPKI for ECDSA public keys
- ECDSA curves: P-256 (prime256v1) and P-384 (secp384r1)
- Output formats: PEM and JWK

# Table of contents

> * [CLI](#cli)
>	- [Creating new keys](#creating-new-keys)
>	- [Converting from one format to another](#converting-from-one-format-to-another)
>	- [Listing OpenID JWK public keys using an OpenID discovery endpoint](#listing-openid-jwk-public-keys-using-an-openid-discovery-endpoint)
>	- [Converting OpenID JWK public keys to PEM files](#converting-openid-jwk-public-keys-to-pem-files)
> * [Using it in Node](#using-it-in-node)

# CLI

For any help with the CLI, use:

```
npx create-keys help
```

## Creating new keys

```
npx create-keys
```

## Converting from one format to another

```
npx create-keys convert private.pem
```

or 

```
npx create-keys convert private.json
```

`create-keys` automatically detectects the format (PEM vs JWK), the type (private vs public keys) and the cipher (RSA vs ECDSA). 

## Listing OpenID JWK public keys using an OpenID discovery endpoint

```
npx create-keys list https://accounts.google.com/.well-known/openid-configuration
```

or 

```
npx create-keys list https://www.googleapis.com/oauth2/v3/certs
```

`create-keys` supports both an OpenID discovery endpoint or the direct `jwks_uri` endpoint. 

## Converting OpenID JWK public keys to PEM files

```
npx create-keys convert https://accounts.google.com/.well-known/openid-configuration
```

or 

```
npx create-keys convert https://www.googleapis.com/oauth2/v3/certs
```

`create-keys` supports both an OpenID discovery endpoint or the direct `jwks_uri` endpoint. 

# Using it in Node

1. Install:
```
npm i create-keys
```
2. In your code:
```js
const { Keypair, Key } = require('create-keys')

const rsaKeypair = new Keypair({ cipher:'rsa', length:1024 })
const ecKeypair = new Keypair({ cipher:'ec', curve:'prime256v1' }) // supported curves: 'prime256v1' and 'secp384r1'

const main = async () => {
	// Creates RSA key pair
	const [rsaPemErrors, rsaPemKeys] = await rsaKeypair.to('pem')
	const [rsaJwkErrors, rsaJwkKeys] = await rsaKeypair.to('jwk')

	// Creates ECDSA key pair
	const [ecPemErrors, ecPemKeys] = await ecKeypair.to('pem')
	const [ecJwkErrors, ecJwkKeys] = await ecKeypair.to('jwk')

	console.log('RSA PRIVATE PEM')
	console.log(rsaPemKeys.private)
	console.log('RSA PUBLIC PEM')
	console.log(rsaPemKeys.public)

	console.log('RSA PRIVATE JWK')
	console.log(JSON.stringify(rsaJwkKeys.private, null, '  '))
	console.log('RSA PUBLIC JWK')
	console.log(JSON.stringify(rsaJwkKeys.public, null, '  '))

	console.log('ECDSA PRIVATE PEM')
	console.log(ecPemKeys.private)
	console.log('ECDSA PUBLIC PEM')
	console.log(ecPemKeys.public)

	console.log('ECDSA PRIVATE JWK')
	console.log(JSON.stringify(ecJwkKeys.private, null, '  '))
	console.log('ECDSA PUBLIC JWK')
	console.log(JSON.stringify(ecJwkKeys.public, null, '  '))

	const [rsaPrivateJwkErros, rsaPrivateJwk] = new Key({ pem:rsaPemKeys.private }).to('jwk')
	console.log('RSA PRIVATE PEM TO JWK')
	console.log(JSON.stringify(rsaPrivateJwk, null, '  '))

	const [ecPublicPemErrors, ecPublicPem] = new Key({ jwk:ecJwkKeys.public }).to('pem')
	console.log('ECDSA PUBLIC JWK TO PEM')
	console.log(ecPublicPem)
}

main()
```





CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/nicolasdao/create-keys/compare/v1.0.0...v1.0.1) (2020-09-29)

## [1.0.0](https://github.com/nicolasdao/create-keys/compare/v0.0.4...v1.0.0) (2020-09-29)


### Features

* Add support for converting files from one format to another, listing OpenID public JWK keys and conveting OpenID JWK keys to PEM ([ea4c95b](https://github.com/nicolasdao/create-keys/commit/ea4c95b186b218b9e3ead76a608cc341b66de056))

### [0.0.4](https://github.com/nicolasdao/create-keys/compare/v0.0.3...v0.0.4) (2020-09-28)

### [0.0.3](https://github.com/nicolasdao/create-keys/compare/v0.0.2...v0.0.3) (2020-09-28)

### 0.0.2 (2020-09-28)


### Features

* Add support for generating RSA and ECDSA pem files and converting them to SSH and JWK format ([c62079a](https://github.com/nicolasdao/create-keys/commit/c62079a7268e93692f0730ccef89111d2ec488cb))
* Add supports for saving keys to files ([1a8246e](https://github.com/nicolasdao/create-keys/commit/1a8246e8b0ed5c5b3a4d39c04f8ca662963830ef))
* Remove support for SSH ([21a7ee0](https://github.com/nicolasdao/create-keys/commit/21a7ee088faf20cef15d16ef2caae016dbfc14f2))


test/index.js
/**
 * Copyright (c) 2020, Cloudless Consulting Pty Ltd.
 * All rights reserved.
 * 
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
*/

// To skip a test, either use 'xit' instead of 'it', or 'describe.skip' instead of 'describe'

const { assert } = require('chai')
const { Keypair, Key } = require('../src')

class WrapperError extends Error {
	constructor(message, stack) {
		super(message)
		this.stack = stack
	}
}

const assertNoErrors = (errors, comment) => {
	try {
		assert.isNotOk(errors, comment)
	} catch(error) {
		const totalErrors = [error, ...errors]
		const stack = totalErrors.map(e => e.stack).join('\n')
		throw new WrapperError(error.message, stack)
	}
}

describe('Keypair', () => {
	describe('constructor', () => {
		it('01 - Should create a RSA pem and JWK keys.', async () => {
			const keypair = new Keypair({ cipher:'rsa' })
			const [pemErrors, pemKeys] = await keypair.to('pem')
			const [jwkErrors, jwkKeys] = await keypair.to('jwk')
			
			assertNoErrors(pemErrors, '01')
			assertNoErrors(jwkErrors, '02')
			assert.isOk(pemKeys, '04')
			assert.isOk(pemKeys.private, '05')
			assert.isOk(pemKeys.public, '06')
			assert.isOk(jwkKeys.private, '07')
			assert.isOk(jwkKeys.public, '08')

		})
		it('02 - Should reconstruct the RSA public pem key from JWK formats.', async () => {
			const keypair = new Keypair({ cipher:'rsa' })
			const [pemErrors, pemKeys] = await keypair.to('pem')
			const [jwkErrors, jwkKeys] = await keypair.to('jwk')

			assertNoErrors(pemErrors, '01')
			assertNoErrors(jwkErrors, '02')

			const key = new Key({ jwk:jwkKeys.public })
			const [pemKeyErrors, pemKey] = key.to('pem')

			assertNoErrors(pemKeyErrors, '03')
			assert.equal(pemKey, pemKeys.public,'04')

		})
		it('03 - Should reconstruct the RSA private pem key from JWK formats.', async () => {
			const keypair = new Keypair({ cipher:'rsa' })
			const [pemErrors, pemKeys] = await keypair.to('pem')
			const [jwkErrors, jwkKeys] = await keypair.to('jwk')
			
			assertNoErrors(pemErrors, '01')
			assertNoErrors(jwkErrors, '02')

			const key = new Key({ jwk:jwkKeys.private })
			const [pemKeyErrors, pemKey] = key.to('pem')

			assertNoErrors(pemKeyErrors, '03')
			assert.equal(pemKey, pemKeys.private,'04')

		})
		it('04 - Should create a ECDSA pem and JWK keys.', async () => {
			const keypair = new Keypair({ cipher:'ec' })
			const [pemErrors, pemKeys] = await keypair.to('pem')
			const [jwkErrors, jwkKeys] = await keypair.to('jwk')
			
			assertNoErrors(pemErrors, '01')
			assertNoErrors(jwkErrors, '02')
			assert.isOk(pemKeys, '04')
			assert.isOk(pemKeys.private, '05')
			assert.isOk(pemKeys.public, '06')
			assert.isOk(jwkKeys.private, '07')
			assert.isOk(jwkKeys.public, '08')

		})
		it('05 - Should reconstruct the ECDSA public pem key from JWK formats.', async () => {
			const keypair = new Keypair({ cipher:'ec' })
			const [pemErrors, pemKeys] = await keypair.to('pem')
			const [jwkErrors, jwkKeys] = await keypair.to('jwk')
			
			assertNoErrors(pemErrors, '01')
			assertNoErrors(jwkErrors, '02')

			const key = new Key({ jwk:jwkKeys.public })
			const [pemKeyErrors, pemKey] = key.to('pem')

			assertNoErrors(pemKeyErrors, '03')
			assert.equal(pemKey, pemKeys.public,'04')

		})
		it('06 - Should reconstruct the ECDSA private pem key from JWK formats.', async () => {
			const keypair = new Keypair({ cipher:'ec' })
			const [pemErrors, pemKeys] = await keypair.to('pem')
			const [jwkErrors, jwkKeys] = await keypair.to('jwk')
			
			assertNoErrors(pemErrors, '01')
			assertNoErrors(jwkErrors, '02')

			const key = new Key({ jwk:jwkKeys.private })
			const [pemKeyErrors, pemKey] = key.to('pem')

			assertNoErrors(pemKeyErrors, '03')
			assert.equal(pemKey, pemKeys.private,'04')

		})
	})	
})

describe('Key', () => {
	describe('#to', () => {
		it('01 - Should convert private key from PEM to JWK format', async () => {
			const keypair = new Keypair({ cipher:'rsa' })
			const [, pemKeys] = await keypair.to('pem')

			const [jwkErrors, jwk] = new Key({ pem:pemKeys.private }).to('jwk')
			
			assertNoErrors(jwkErrors, '01')
			assert.isOk(jwk, '02')
			assert.equal(jwk.kty, 'RSA', '03')
			assert.isOk(jwk.e, '04')
			assert.isOk(jwk.n, '05')
			assert.isOk(jwk.p, '06')
			assert.isOk(jwk.q, '07')
			assert.isOk(jwk.d, '08')
			assert.isOk(jwk.dp, '09')
			assert.isOk(jwk.dq, '10')
			assert.isOk(jwk.qi, '11')
			assert.isNotOk(jwk.dmp1, '12')
			assert.isNotOk(jwk.dmq1, '13')
			assert.isNotOk(jwk.coeff, '14')
		})
	})
})






test/utils.js
/**
 * Copyright (c) 2020, Cloudless Consulting Pty Ltd.
 * All rights reserved.
 * 
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
*/

// To skip a test, either use 'xit' instead of 'it', or 'describe.skip' instead of 'describe'

const { assert } = require('chai')
const { numberToBase64, base64ToNumber } = require('../src/utils')

describe('utils', () => {
	describe('#numberToBase64', () => {
		it('01 - Should convert big numbers into base64', () => {
			const b64 = numberToBase64(65537)
			assert.equal(b64, 'AQAB', '01')
		})
	})
	describe('#base64ToNumber', () => {
		it('01 - Should convert base64 to big number', () => {
			const nbr = base64ToNumber('AQAB')
			assert.equal(nbr, 65537, '01')
		})
	})
})

src/rsa.js
const getPem = require('rsa-pem-from-mod-exp')
const { error: { catchErrors, wrapErrors } } = require('puffy')
// const { pem2jwk } = require('pem-jwk')
const NodeRSA = require('node-rsa')
const { jwtBuffer, jwtB64 } = require('./utils')

/**
 * Create a PEM string from a public RSA modulus and exponent. 
 * 
 * @param  {String} input.modulus || input.n
 * @param  {String} input.exponent || input.e
 * 
 * @return {String} publicKey
 */
const jwkToPem = (input={}) => catchErrors(() => {
	const errorMsg = 'Failed to create public RSA key from modulus and exponent'
	const { modulus, n, exponent, e } = input
	const mod = modulus || n
	const exp = exponent || e
	if (!mod)
		throw new Error(`${errorMsg}. Missing required 'modulus|n'`)
	if (!exp)
		throw new Error(`${errorMsg}. Missing required 'exponent|e'`)

	const jwk = modulus && exponent ? { kty:'RSA', n:modulus, e:exponent } : input

	const isPrivate = jwk.d !== undefined
	
	if (isPrivate) {
		const jwkBuff = jwtBuffer(jwk)
		try {
			const key = new NodeRSA()
			key.importKey(jwkBuff, 'components')
			return key.exportKey('pkcs8-private-pem') + '\n'
		} catch(err) {
			throw wrapErrors(errorMsg, [err])
		}
	} else {
		const jwkBase64 = jwtB64(jwk)
		try {
			return getPem(jwkBase64.n, jwkBase64.e)
		} catch(err) {
			throw wrapErrors(errorMsg, [err])
		}
	}
})

const pemToJwk = (pemKey, isPrivate) => catchErrors(() => {
	const errorMsg = 'Failed to convert RSA key from PEM to JWK format'

	try {
		const rsaKey = new NodeRSA(pemKey)
		const jwt = rsaKey.exportKey(`components-${isPrivate ? 'private' : 'public'}-pem`)
		const core = {
			kty: 'RSA',
			e: jwt.e,
			n: jwt.n
		}
		if (isPrivate) 
			return jwtB64({
				...core,
				d: jwt.d,
				p: jwt.p,
				q: jwt.q,
				dp: jwt.dmp1,
				dq: jwt.dmq1,
				qi: jwt.coeff
			})
		else
			return jwtB64(core)
	} catch(err) {
		throw wrapErrors(errorMsg, [err])
	}
})

module.exports = {
	jwkToPem,
	pemToJwk
}


src/ec-key.js
const crypto = require('crypto')
const asn = require('asn1.js')
const util = require('util')

/* ========================================================================== *
 * From RFC-4492 (Appendix A) Equivalent Curves (Informative)                 *
 * ========================================================================== *
 *                                                                            *
 * +------------------------------------------------------------------------+ *
 * |                         Curve names chosen by                          | *
 * |                   different standards organizations                    | *
 * +-----------+------------+------------+------------+---------------------+ *
 * |   SECG    | ANSI X9.62 |    NIST    |  OpenSSL   |      ASN.1 OID      | *
 * +-----------+------------+------------+------------+---------------------+ *
 * | secp256r1 | prime256v1 | NIST P-256 | prime256v1 | 1.2.840.10045.3.1.7 | *
 * | secp256k1 |            |            | secp256k1  | 1.3.132.0.10        | *
 * | secp384r1 |            | NIST P-384 | secp384r1  | 1.3.132.0.34        | *
 * | secp521r1 |            | NIST P-521 | secp521r1  | 1.3.132.0.35        | *
 * +-----------+------------+------------+------------+---------------------+ *
 * ========================================================================== */

/* Byte lengths for validation */
const lengths = {
	prime256v1 : Math.ceil(256 / 8),
	secp256k1 : Math.ceil(256 / 8),
	secp384r1 : Math.ceil(384 / 8),
	secp521r1 : Math.ceil(521 / 8)
}

/* JWK curve names */
const jwkCurves = {
	prime256v1 : 'P-256',
	secp256k1 : 'P-256K', /* NOT A STANDARD NAME: See the README.md file */
	secp384r1 : 'P-384',
	secp521r1 : 'P-521'
}

/* OpenSSL curve names */
const curves = {
	'P-256'  : 'prime256v1',
	'P-256K' : 'secp256k1', /* NOT A STANDARD NAME: See the README.md file */
	'P-384'  : 'secp384r1',
	'P-521'  : 'secp521r1'
}

/* ========================================================================== *
 * ASN.1                                                                      *
 * ========================================================================== */

const ASN1ECRfc5915Key = asn.define('Rfc5915Key', function() {
	this.seq().obj(
		this.key('version').int(),
		this.key('privateKey').octstr(),
		this.key('parameters').optional().explicit(0).objid({
			'1 2 840 10045 3 1 7' : 'prime256v1',
			'1 3 132 0 10'        : 'secp256k1',
			'1 3 132 0 34'        : 'secp384r1',
			'1 3 132 0 35'        : 'secp521r1'
		}),
		this.key('publicKey').optional().explicit(1).bitstr()
	)
})

/* ========================================================================== */

const ASN1ECPkcs8Key = asn.define('Pkcs8Key', function() {
	this.seq().obj(
		this.key('version').int(),
		this.key('algorithmIdentifier').seq().obj(
			this.key('privateKeyType').objid({
				'1 2 840 10045 2 1': 'EC'
			}),
			this.key('parameters').objid({
				'1 2 840 10045 3 1 7' : 'prime256v1',
				'1 3 132 0 10'        : 'secp256k1',
				'1 3 132 0 34'        : 'secp384r1',
				'1 3 132 0 35'        : 'secp521r1'
			})
		),
		this.key('privateKey').octstr()
	)
})

/* ========================================================================== */

const ASN1ECSpkiKey = asn.define('SpkiKey', function() {
	this.seq().obj(
		this.key('algorithmIdentifier').seq().obj(
			this.key('publicKeyType').objid({
				'1 2 840 10045 2 1': 'EC'
			}),
			this.key('parameters').objid({
				'1 2 840 10045 3 1 7' : 'prime256v1',
				'1 3 132 0 10'        : 'secp256k1',
				'1 3 132 0 34'        : 'secp384r1',
				'1 3 132 0 35'        : 'secp521r1'
			})
		),
		this.key('publicKey').bitstr()
	)
})

/* ========================================================================== *
 * ASN.1 PARSING                                                              *
 * ========================================================================== */

/* Parse a public key buffer, split X and Y */
function parsePublicKeyBuffer(curve, buffer) {
	var bytes = lengths[curve]
	if (buffer[0] == 4) {
		if (buffer.length != ((bytes * 2) + 1)) throw new TypeError('Invalid uncompressed key size')
		return {
			c: curve,
			x: buffer.slice(1, bytes + 1),
			y: buffer.slice(bytes + 1),
		}
	} else {
		throw new TypeError('Compressed key unsupported')
	}
}

/* Parse PKCS8 from RFC 5208 */
function parsePkcs8(buffer) {
	var key = ASN1ECPkcs8Key.decode(buffer, 'der')
	var privateKeyWrapper = ASN1ECRfc5915Key.decode(key.privateKey, 'der')
	var curve = key.algorithmIdentifier.parameters
	var bytes = lengths[curve]

	var privateKey = privateKeyWrapper.privateKey
	if (privateKey.length > bytes) {
		throw new TypeError('Invalid private key size: expected ' + bytes + ' gotten ' + privateKey.length)
	}

	var components = parsePublicKeyBuffer(curve, privateKeyWrapper.publicKey.data)
	components.d = privateKey
	return components
}

/* Parse EC from RFC 5915 */
function parseRfc5915(buffer) {
	var key = ASN1ECRfc5915Key.decode(buffer, 'der')
	var bytes = lengths[key.parameters]

	var privateKey = key.privateKey
	if (privateKey.length > bytes) {
		throw new TypeError('Invalid private key size: expected ' + bytes + ' gotten ' + privateKey.length)
	}

	var components = parsePublicKeyBuffer(key.parameters, key.publicKey.data)
	components.d = privateKey
	return components
}

/* Parse SPKI from RFC 5280 */
function parseSpki(buffer) {
	var key = ASN1ECSpkiKey.decode(buffer, 'der')
	return parsePublicKeyBuffer(key.algorithmIdentifier.parameters, key.publicKey.data)
}

/* ========================================================================== *
 * PEM PARSING                                                                *
 * ========================================================================== */

const pemRfc5915RE = /-+BEGIN EC PRIVATE KEY-+([\s\S]+)-+END EC PRIVATE KEY-+/m
const pemPkcs8RE   = /-+BEGIN PRIVATE KEY-+([\s\S]*)-+END PRIVATE KEY-+/m
const pemSpkiRE    = /-+BEGIN PUBLIC KEY-+([\s\S]*)-+END PUBLIC KEY-+/m

function parsePem(pem) {
	if (! util.isString(pem)) throw new TypeError('PEM must be a string')

	var match = pem.match(pemRfc5915RE)
	
	if (match)
		return parseRfc5915(Buffer.from(match[1].replace(/[\s-]/mg, ''), 'base64'))
	match = pem.match(pemPkcs8RE)
	if (match)
		return parsePkcs8(Buffer.from(match[1].replace(/[\s-]/mg, ''), 'base64'))
	match = pem.match(pemSpkiRE)
	if (match)
		return parseSpki(Buffer.from(match[1].replace(/[\s-]/mg, ''), 'base64'))
	
	throw new TypeError('Unrecognized PEM key structure')
}

/* ========================================================================== *
 * CLASS DEFINITION                                                           *
 * ========================================================================== */

function ECKey(key, format) {
	if (!(this instanceof ECKey)) return new ECKey(key, format)

	var curve, d, x, y

	if (! format) format = 'pem'

	// BUFFER KEYS: either in "pkcs8" or "spki" format (base64) or "pem" (ascii)
	if (util.isBuffer(key)) {

		if (format == 'pem') {
			const k = parsePem(key.toString('ascii'))
			curve = k.c
			x = k.x
			y = k.y
			d = k.d

		} else if ((format == 'pkcs8') || (format == 'rfc5208')) {
			const k = parsePkcs8(key)
			curve = k.c
			x = k.x
			y = k.y
			d = k.d


		} else if ((format == 'spki') || (format == 'rfc5280')) {
			const k = parseSpki(key)
			curve = k.c
			x = k.x
			y = k.y
			d = k.d

		} else {
			throw new TypeError('Unknown format for EC Key "' + format + '"')
		}
	}

	// STRING KEYS: base64 all the time, but also allowed in PEM
	else if (util.isString(key)) {

		if (format == 'pem') {
			const k = parsePem(key)
			curve = k.c
			x = k.x
			y = k.y
			d = k.d

		} else if ((format == 'pkcs8') || (format == 'rfc5208')) {
			const k = parsePkcs8(Buffer.from(key, 'base64'))
			curve = k.c
			x = k.x
			y = k.y
			d = k.d


		} else if ((format == 'spki') || (format == 'rfc5280')) {
			const k = parseSpki(Buffer.from(key, 'base64'))
			curve = k.c
			x = k.x
			y = k.y
			d = k.d

		} else {
			throw new TypeError('Unknown format for EC Key "' + format + '"')
		}
	}

	// OBJECT KEY: needs to contain "(curve|crv)", "(d|privateKey)" if it's a
	// private key and "(publicKey|x,y)" always required (both for priv and pub)
	else if (util.isObject(key)) {

		// Curves
		if (util.isString(key.curve)) {
			curve = key.curve
		} else if (util.isString(key.crv)) {
			curve = curves[key.crv] || key.crv
		}

		// Private key or "d"
		if (util.isBuffer(key.privateKey)) {
			d = key.privateKey
		} else if (util.isString(key.privateKey)) {
			d = Buffer.from(key.privateKey, 'base64')
		} else if (util.isBuffer(key.d)) {
			d = key.d
		} else if (util.isString(key.d)) {
			d = Buffer.from(key.d, 'base64')
		}

		// Public key, or x and y
		if (util.isBuffer(key.publicKey)) {
			const k = parsePublicKeyBuffer(curve, key.publicKey)
			x = k.x
			y = k.y

		} else if (util.isString(key.publicKey)) {
			const k = parsePublicKeyBuffer(curve, Buffer.from(key.publicKey, 'base64'))
			x = k.x
			y = k.y

		} else {
			// Need to get x and y
			if (util.isBuffer(key.x)) {
				x = key.x
			} else if (util.isString(key.x)) {
				x = Buffer.from(key.x, 'base64')
			}

			if (util.isBuffer(key.y)) {
				y = key.y
			} else if (util.isString(key.y)) {
				y = Buffer.from(key.y, 'base64')
			}
		}

	} else {
		throw new TypeError('Unrecognized format for EC key')
	}

	// Validate curve, d, x and y
	if (! curve) throw new TypeError('EC Key curve not specified')
	if ((! x) || (! y)) throw new TypeError('Public EC Key point unavailable')

	var length = lengths[curve]
	if (! length) throw new TypeError('EC Key curve "' + curve + '" invalid')
	if (x.length != length) throw new TypeError('Public EC Key point X of wrong length')
	if (y.length != length) throw new TypeError('Public EC Key point Y of wrong length')
	if (d && (y.length != length)) throw new TypeError('Private EC Key of wrong length')

	// Define our properties
	Object.defineProperties(this, {
		'curve': {
			enumerable: true,
			configurable: false,
			value: curve
		},
		'isPrivateECKey': {
			enumerable: true,
			configurable: false,
			value: (d != null)
		},
		'x': {
			enumerable: true,
			configurable: false,
			get: function() {
				return Buffer.from(x)
			}
		},
		'y': {
			enumerable: true,
			configurable: false,
			get: function() {
				return Buffer.from(y)
			}
		},

		// "non-enumerable"
		'jsonCurve': {
			enumerable: false,
			configurable: false,
			value: jwkCurves[curve]
		},
		'publicCodePoint': {
			enumerable: false,
			configurable: false,
			get: function() {
				return Buffer.concat([Buffer.from([0x04]), x, y])
			}
		}
	})

	// The "d" (private key) is optional
	if (d) Object.defineProperty(this, 'd', {
		enumerable: true,
		configurable: false,
		get: function() {
			return Buffer.from(d)
		}
	})
}

ECKey.createECKey = function(curve) {
	if (!curve) curve = 'prime256v1'
	if (curves[curve]) curve = curves[curve]
	if (!jwkCurves[curve]) throw new TypeError('Invalid/unknown curve "' + curve + '"')

	var ecdh = crypto.createECDH(curve)

	ecdh.generateKeys()
	return new ECKey({
		privateKey: ecdh.getPrivateKey(),
		publicKey: ecdh.getPublicKey(),
		curve: curve
	})
}

/* ========================================================================== *
 * ECDH/SIGNING/VALIDATION                                                    *
 * ========================================================================== */

ECKey.prototype.computeSecret = function computeSecret(key) {
	return this.createECDH().computeSecret(key)
}

ECKey.prototype.createECDH = function createECDH() {
	if (this.isPrivateECKey) {
		var ecdh = crypto.createECDH(this.curve)
		ecdh.setPublicKey(this.publicCodePoint)
		ecdh.setPrivateKey(this.d)

		var ecdhComputeSecret = ecdh.computeSecret
		ecdh.computeSecret = function(key) {
			if (key instanceof ECKey) {
				return ecdhComputeSecret.call(ecdh, key.publicCodePoint)
			} else {
				return ecdhComputeSecret.apply(ecdh, arguments)
			}
		}
		return ecdh
	} else {
		throw new Error('Can only create ECDH from private keys')
	}
}

ECKey.prototype.createSign = function createSign(hash) {
	if (! this.isPrivateECKey) throw new Error('EC Private Key needed to sign')
	var sign = crypto.createSign('RSA-' + hash) // RSA works with EC keys, too
	var signFunction = sign.sign
	var self = this
	sign.sign = function(format) {
		return signFunction.call(sign, self.toString('pem'), format)
	}
	return sign
}

ECKey.prototype.createVerify = function createVerify(hash) {
	var verify = crypto.createVerify('RSA-' + hash) // RSA works with EC keys, too
	var verifyFunction = verify.verify
	var key = this.isPrivateECKey ? this.asPublicECKey() : this
	verify.verify = function(signature, format) {
		return verifyFunction.call(verify, key.toString('pem'), signature, format)
	}
	return verify
}

ECKey.prototype.asPublicECKey = function asPublicECKey() {
	if (! this.isPrivateECKey) return this
	return new ECKey({
		curve: this.curve,
		x: this.x,
		y: this.y
	})
}


/* ========================================================================== *
 * CONVERSION                                                                 *
 * ========================================================================== */

ECKey.prototype.toBuffer = function toBuffer(format) {
	if (! format) format = 'pem'

	// Simple PEM conversion, wrapping the string in the buffer
	if (format == 'pem') return Buffer.from(this.toString('pem'), 'ascii')

	if (this.isPrivateECKey) {
		// Strip leading zeroes from private key
		var d = this.d
		while (d[0] == 0) d = d.slice(1)

		// Known formats: "pkcs8" (default), "pem", "openssl"
		if ((format == 'pkcs8') || (format == 'rfc5208')) {

			// Encode in PKCS8
			return ASN1ECPkcs8Key.encode({
				version: 0,
				algorithmIdentifier: {
					privateKeyType: 'EC',
					parameters: this.curve,
				},
				// Private key is RFC5915 minus curve
				privateKey: ASN1ECRfc5915Key.encode({
					version: 1,
					privateKey: d,
					publicKey: { data: this.publicCodePoint }
				}, 'der')
			}, 'der')

		} else if (format == 'rfc5915') {

			// Simply encode in ASN.1
			return ASN1ECRfc5915Key.encode({
				version: 1,
				privateKey: d,
				parameters: this.curve,
				publicKey: { data: this.publicCodePoint }
			}, 'der')

		} else {
			throw new TypeError('Unknown format for private key "' + format + '"')
		}

	} else {

		if ((format == 'spki') || (format == 'rfc5280')) {
			return ASN1ECSpkiKey.encode({
				algorithmIdentifier: {
					publicKeyType: 'EC',
					parameters: this.curve
				},
				publicKey: { data: this.publicCodePoint }
			}, 'der')

		} else {
			throw new TypeError('Unknown format for public key "' + format + '"')
		}
	}
}

ECKey.prototype.toString = function toString(format) {
	if (! format) format = 'pem'

	if (this.isPrivateECKey) {
		if (format == 'pem') { // pkcs8, wrapped
			return '-----BEGIN PRIVATE KEY-----\n'
           + this.toBuffer('pkcs8').toString('base64').match(/.{1,64}/g).join('\n')
           + '\n-----END PRIVATE KEY-----\n'

		} else if (format == 'rfc5915') { // rfc5915, wrapped
			return '-----BEGIN EC PRIVATE KEY-----\n'
           + this.toBuffer('rfc5915').toString('base64').match(/.{1,64}/g).join('\n')
           + '\n-----END EC PRIVATE KEY-----\n'

		} else if ((format == 'pkcs8') || (format == 'rfc5208')) {
			return this.toBuffer('pkcs8').toString('base64')

		} else if ((format == 'spki') || (format == 'rfc5280')) {
			return this.toBuffer('spki').toString('base64')

		} else {
			throw new TypeError('Unknown format for private key "' + format + '"')
		}

	} else {
		if (format == 'pem') {
			return '-----BEGIN PUBLIC KEY-----\n'
           + this.toBuffer('spki').toString('base64').match(/.{1,64}/g).join('\n')
           + '\n-----END PUBLIC KEY-----\n'

		} else if ((format == 'spki') || (format == 'rfc5280')) {
			return this.toBuffer('spki').toString('base64')

		} else {
			throw new TypeError('Unknown format for public key "' + format + '"')
		}
	}
}

ECKey.prototype.toJSON = function toJSON() {
	function urlsafe(buffer) {
		return buffer.toString('base64')
			.replace(/\+/g, '-')
			.replace(/\//g, '_')
			.replace(/=/g,   '')
	}

	var jwk = {
		kty: 'EC',
		crv: jwkCurves[this.curve],
		x: urlsafe(this.x),
		y: urlsafe(this.y),
	}

	var d = this.d
	if (d) {
		var bytes = lengths[this.curve]
		if (d.length < bytes) {
			var remaining = bytes - d.length
			d = Buffer.concat([Buffer.from(remaining).fill(0), d])
		}
		jwk.d = urlsafe(d)
	}

	return jwk
}


/* ========================================================================== *
 * EXPORTS                                                                    *
 * ========================================================================== */

exports = module.exports = ECKey

